usr/src/ftpd-diku/support/authuser.3
.TH authuser 3
.SH NAME
authuser \- library to get information from a remote Authentication Server
.SH SYNTAX
.B #include <authuser.h>
.PP
.B unsigned short auth_tcpport;
.PP
.B char *auth_xline(user,fd,&in);
.PP
.B int auth_fd(fd,&in,&local,&remote);
.PP
.B char *auth_tcpuser(in,local,remote);
.PP
.B char *user;
.br
.B int fd;
.br
.B unsigned long in;
.br
.B unsigned short local;
.br
.B unsigned short remote;
.SH DESCRIPTION
The
.I authuser
library provides a simple interface for
finding out the remote identity
of a connection through the
Authentication Server
as specified by RFC 931.
Use the -lauthuser loader option
to compile a program with this library.
.PP
.B auth_xline(user,fd,&in)
returns a line of the form X-Auth-User: user or X-Forgery-By: username,
depending upon what the host on the other side of
.B fd
thinks of the user.
This is particularly appropriate for
mail and news headers.
.PP
If the remote host reports that
.B user
owns the connection on that side,
.B auth_xline
will return X-Auth-User: user.
If the remote host reports that a different
.B username
owns the connection,
.B auth_xline
will return X-Forgery-By: username.
If user is NULL,
it returns X-Auth-User: username
with the username reported by the remote host.
If 
.B fd
is not a TCP connection
or authentication is impossible,
.B auth_xline
returns NULL, setting errno appropriately.
.PP
The line is not cr-lf terminated.
It is stored in a static area
which is overwritten on each call to
.B auth_xline.
.B auth_xline
places the Internet address of the other host into in.
.PP
.B auth_fd(fd,&in,&local,&remote)
retrieves address information from the connection in socket
.B fd.
It places the
Internet address of the host on other side into
.B in
and the local and remote
TCP ports into 
.B local
and 
.B remote.
.B auth_fd
returns -1 upon error, setting errno appropriately.
.PP
.B auth_tcpuser(in,local,remote)
returns the name of the user on the other end of the TCP connection
between
.B remote@in
and
.B local.
If authentication is impossible,
.B auth_tcpuser
returns
NULL, setting errno appropriately.
The user name is stored in a static area
which is overwritten on each call to
.B auth_tcpuser
and
.B auth_xline.
.PP
The authentication routines check with the
remote Authentication Server on port
.B auth_tcpport,
which defaults to 113
as specified by RFC 931.
You can set
.B auth_tcpport
to other values
for nonstandard implementations.
.PP
.SH RESTRICTIONS
.I authuser
does no backslash interpretation
upon the remote user name.
Hopefully the next revision of RFC 931
will make clear exactly what backslash
interpretation should be going on.
.PP
.I authuser
does not use the operating system type
information provided by the Authentication Server.
.SH VERSION
authuser version 3.1, May 6, 1991.
.SH AUTHOR
Placed into the public domain by Daniel J. Bernstein.
.SH REFERENCES
The authentication server is more secure than passwords
in some ways, but less secure than passwords in many ways.
(It's certainly better than no password at all---e.g., for
mail or news.)
It is not the final solution.
For an excellent discussion of security problems within
the TCP/IP protocol suite, see
Steve Bellovin's article
``Security Problems in the TCP/IP Protocol Suite.''
.SH "SEE ALSO"
authtcp(1),
attachport(1),
getpeername(3),
getsockname(3),
tcp(4),
authd(8)


usr/src/ftpd-diku/support/authuser.c
/*
5/6/91 DJB baseline authuser 3.1. Public domain.
*/

#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <errno.h>
#include <ctype.h>
extern int errno;
#include "authuser.h"

unsigned short auth_tcpport = 113;

#define SIZ 500 /* various buffers */

static int usercmp(u,v)
register char *u;
register char *v;
{
 /* is it correct to consider Foo and fOo the same user? yes */
 /* but the function of this routine may change later */
 while (*u && *v)
   if (tolower(*u) != tolower(*v))
     return tolower(*u) - tolower(*v);
   else
     ++u,++v;
 return *u || *v;
}

static char authline[SIZ];

char *auth_xline(user,fd,in)
register char *user; /* the supposed name of the user, NULL if unknown */
register int fd; /* the file descriptor of the connection */
register unsigned long *in;
{
 unsigned short local;
 unsigned short remote;
 register char *ruser;

 if (auth_fd(fd,in,&local,&remote) == -1)
   return 0;
 ruser = auth_tcpuser(*in,local,remote);
 if (!ruser)
   return 0;
 if (!user)
   user = ruser; /* forces X-Auth-User */
 (void) sprintf(authline,
	 (usercmp(ruser,user) ? "X-Forgery-By: %s" : "X-Auth-User: %s"),
	 ruser);
 return authline;
}

int auth_fd(fd,in,local,remote)
register int fd;
register unsigned long *in;
register unsigned short *local;
register unsigned short *remote;
{
 struct sockaddr_in sa;
 int dummy;

 dummy = sizeof(sa);
 if (getsockname(fd,&sa,&dummy) == -1)
   return -1;
 if (sa.sin_family != AF_INET)
  {
   errno = EAFNOSUPPORT;
   return -1;
  }
 *local = ntohs(sa.sin_port);
 dummy = sizeof(sa);
 if (getpeername(fd,&sa,&dummy) == -1)
   return -1;
 *remote = ntohs(sa.sin_port);
 *in = sa.sin_addr.s_addr;
 return 0;
}

static char ruser[SIZ];
static char realbuf[SIZ];
static char *buf;

char *auth_tcpuser(in,local,remote)
register unsigned long in;
register unsigned short local;
register unsigned short remote;
{
 struct sockaddr_in sa;
 register int s;
 register int buflen;
 register int w;
 register int saveerrno;
 char ch;
 unsigned short rlocal;
 unsigned short rremote;

 if ((s = socket(AF_INET,SOCK_STREAM,0)) == -1)
   return 0;
 sa.sin_family = AF_INET;
 sa.sin_port = htons(auth_tcpport);
 sa.sin_addr.s_addr = in;
 if (connect(s,&sa,sizeof(sa)) == -1)
  {
   saveerrno = errno;
   (void) close(s);
   errno = saveerrno;
   return 0;
  }

 buf = realbuf;
 (void) sprintf(buf,"%u , %u\r\n",(unsigned int) remote,(unsigned int) local);
 /* note the reversed order---the example in the RFC is misleading */
 buflen = strlen(buf);
 while ((w = write(s,buf,buflen)) < buflen)
   if (w == -1) /* should we worry about 0 as well? */
    {
     saveerrno = errno;
     (void) close(s);
     errno = saveerrno;
     return 0;
    }
   else
    {
     buf += w;
     buflen -= w;
    }
 buf = realbuf;
 while ((w = read(s,&ch,1)) == 1)
  {
   *buf = ch;
   if ((ch != ' ') && (ch != '\t') && (ch != '\r'))
     ++buf;
   if ((buf - realbuf == sizeof(realbuf) - 1) || (ch == '\n'))
     break;
  }
 if (w == -1)
  {
   saveerrno = errno;
   (void) close(s);
   errno = saveerrno;
   return 0;
  }
 *buf = '\0';

 if (sscanf(realbuf,"%hd,%hd: USERID :%*[^:]:%s",&rremote,&rlocal,ruser) < 3)
  {
   (void) close(s);
   errno = EIO;
   /* makes sense, right? well, not when USERID failed to match ERROR */
   /* but there's no good error to return in that case */
   return 0;
  }
 if ((remote != rremote) || (local != rlocal))
  {
   (void) close(s);
   errno = EIO;
   return 0;
  }
 /* XXX: we're not going to do any backslash processing */
 (void) close(s);
 return ruser;
}
